arangod/Aql/Ast.cpp

```diff
@@ -636,63 +712,39 @@ AstNode* Ast::createNodeDataSource(arangodb::CollectionNameResolver const& resol
-    // case it is ok that the datasource does not exist, but we need to track
-    // the names of datasources used in the query
-    _query->collections()->add(nameString, accessType);
+  LogicalCollection::Category const* category = LogicalCollection::category();
+  StringRef nameRef;
+  std::tie(category, nameRef) = injectDataSourceInQuery(*_query, resolver, accessType, failIfDoesNotExist, nameString);
 
+  if (category == LogicalCollection::category()) {
     return createNodeCollectionNoValidation(name, nameLength, nameString, accessType);
-  }
-            
-  // query actual name from datasource... this may be different to the
-  // name passed into this function, because the user may have accessed
-  // the collection by its numeric id
-  auto const& dataSourceName = dataSource->name();
-  name = _query->registerString(dataSourceName.data(), dataSourceName.size());
-  
-  // add datasource to query
-  _query->collections()->add(dataSourceName, accessType);
-  
-  if (dataSource->category() == LogicalCollection::category()) {
-    // it's a collection!
-    return createNodeCollectionNoValidation(name, dataSourceName.size(), dataSourceName, accessType);
-  }
-
-  if (dataSource->category() == LogicalView::category()) {
-    // it's a view!
+  } else if (category == LogicalView::category()) {
     AstNode* node = createNode(NODE_TYPE_VIEW);
-    node->setStringValue(name, dataSourceName.size());
-    _query->addView(dataSourceName);
-
+    node->setStringValue(nameRef.data(), nameRef.size());
     return node;
   }
-
-  THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, "unexpected datasource type");   
+  // injectDataSourceInQuery is supposed to throw in this case.
+  TRI_ASSERT(false);
 }
 
 /// @brief create an AST collection node
-AstNode* Ast::createNodeCollection(char const* name,
+AstNode* Ast::createNodeCollection(arangodb::CollectionNameResolver const& resolver,
+                                   char const* name,
                                    size_t nameLength,
                                    AccessMode::Type accessType) {
   std::string const nameString = validateDataSourceName(name, nameLength, true);
+  LogicalCollection::Category const* category = LogicalCollection::category();
+  StringRef nameRef;
+  // namestring might be empty
+  std::tie(category, nameRef) = injectDataSourceInQuery(*_query, resolver, accessType, false, nameString);
+  if (category == LogicalCollection::category()) {
+    // add collection to query
+    _query->addCollection(nameRef, accessType);
 
-  // add collection to query
-  _query->collections()->add(nameString, accessType);
-
-  // call private function after validation
-  return createNodeCollectionNoValidation(name, nameLength, nameString, accessType);
+    // call private function after validation
+    return createNodeCollectionNoValidation(nameRef.data(), nameRef.size(), nameRef.toString(), accessType);
```

Comment:
this call (to `createNodeCollectionNoValidation`) creates an AstNode of type `NODE_TYPE_COLLECTION` which contains a (non-owning) pointer to the collection name. The name pointed to is the same as `nameRef.data()`, which may be a heap-allocated string owned by the AQL query, or, worse, a pointer into the stack-allocated std::string `nameString`, which is passed as last parameter into `injectDataSourceInQuery`. `injectDataSourceInQuery` may return a StringRef pointing to just this stack-allocated string in its first return statement.